In [9]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [12]:
!pip3 install xlrd

In [13]:
data = pd.read_excel(r"Concrete_Data.xls")
data.head()

,Cement (component 1)(kg in a m^3 mixture),Blast Furnace Slag (component 2)(kg in a m^3 mixture),Fly Ash (component 3)(kg in a m^3 mixture),Water (component 4)(kg in a m^3 mixture),Superplasticizer (component 5)(kg in a m^3 mixture),Coarse Aggregate (component 6)(kg in a m^3 mixture),Fine Aggregate (component 7)(kg in a m^3 mixture),Age (day),"Concrete compressive strength(MPa, megapascals)"
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28,79.986111
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28,61.887366
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270,40.269535
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365,41.052780
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360,44.296075


In [14]:
data.columns

Index(['Cement (component 1)(kg in a m^3 mixture)',
       'Blast Furnace Slag (component 2)(kg in a m^3 mixture)',
       'Fly Ash (component 3)(kg in a m^3 mixture)',
       'Water  (component 4)(kg in a m^3 mixture)',
       'Superplasticizer (component 5)(kg in a m^3 mixture)',
       'Coarse Aggregate  (component 6)(kg in a m^3 mixture)',
       'Fine Aggregate (component 7)(kg in a m^3 mixture)', 'Age (day)',
       'Concrete compressive strength(MPa, megapascals) '],
      dtype='object')

In [16]:
data.columns = data.columns.str.strip()

In [18]:
data.columns

Index(['Cement (component 1)(kg in a m^3 mixture)',
       'Blast Furnace Slag (component 2)(kg in a m^3 mixture)',
       'Fly Ash (component 3)(kg in a m^3 mixture)',
       'Water  (component 4)(kg in a m^3 mixture)',
       'Superplasticizer (component 5)(kg in a m^3 mixture)',
       'Coarse Aggregate  (component 6)(kg in a m^3 mixture)',
       'Fine Aggregate (component 7)(kg in a m^3 mixture)', 'Age (day)',
       'Concrete compressive strength(MPa, megapascals)'],
      dtype='object')

Objective: To come up with a machine learning model that can predict the concrete compressive strength

In [19]:
data.dtypes

Cement (component 1)(kg in a m^3 mixture)                float64
Blast Furnace Slag (component 2)(kg in a m^3 mixture)    float64
Fly Ash (component 3)(kg in a m^3 mixture)               float64
Water  (component 4)(kg in a m^3 mixture)                float64
Superplasticizer (component 5)(kg in a m^3 mixture)      float64
Coarse Aggregate  (component 6)(kg in a m^3 mixture)     float64
Fine Aggregate (component 7)(kg in a m^3 mixture)        float64
Age (day)                                                  int64
Concrete compressive strength(MPa, megapascals)          float64
dtype: object

In [20]:
data.shape

(1030, 9)

In [24]:
from sklearn.model_selection import train_test_split
ml_data = data.copy()
x = ml_data.drop(columns=["Concrete compressive strength(MPa, megapascals)"])
y = ml_data["Concrete compressive strength(MPa, megapascals)"]

## Training set and testsing set
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
print(f"training x: {x_train.shape} y: {y_train.shape}")
print(f"testing x: {x_test.shape} y: {y_test.shape}")

training x: (824, 8) y: (824,)
testing x: (206, 8) y: (206,)


 Training a linear regression model

In [23]:
from sklearn.linear_model import LinearRegression

lr = LinearRegression()
lr.fit(x_train, y_train)
y_predictions = lr.predict(x_test)

Model Evaluation

In [25]:
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error
r2 = r2_score(y_test, y_predictions)
mse = mean_squared_error(y_test, y_predictions)
mae = mean_absolute_error(y_test, y_predictions)

print(f"R^2: {r2} \n MSE: {mse} \n MAE: {mae}")

R^2: 0.627541605542902 
 MSE: 95.97548435337708 
 MAE: 7.745392872421345


- R^2: 0.627541605542902: this shos that our model is 62% accurate
- MSE: 95.97548435337708: 
- MAE: 7.745392872421345: This tells us that the value our model is likely to predict will be 7.7454 up or down the actual value

Training a decison tree regression model

In [33]:
from sklearn.tree import DecisionTreeRegressor
dtr = DecisionTreeRegressor(criterion="squared_error", max_depth=2)
dtr.fit(x_train, y_train)
y_predictions_dtr = dtr.predict(x_test)

In [34]:
r2 = r2_score(y_test, y_predictions)
mse = mean_squared_error(y_test, y_predictions)
mae = mean_absolute_error(y_test, y_predictions)

print(f"R^2: {r2} \n MSE: {mse} \n MAE: {mae}")

R^2: 0.627541605542902 
 MSE: 95.97548435337708 
 MAE: 7.745392872421345
